File: server/gcpConfig.ts
/**
 * Google Cloud Platform Client Initialization
 * 
 * Centralized initialization of all GCP clients:
 * - Cloud Storage (GCS)
 * - Cloud KMS
 * - Healthcare API (FHIR, DICOM, NLP)
 * - Document AI
 * 
 * Authentication via GOOGLE_APPLICATION_CREDENTIALS or Application Default Credentials
 */

import { Storage } from "@google-cloud/storage";
import { KeyManagementServiceClient } from "@google-cloud/kms";
import { GCP_CONFIG, isGCPConfigured, isKMSConfigured } from "./config/gcpConstants";

function parseGCSCredentials(): { credentials: any; projectId: string } | null {
  let credentialsJson = process.env.GCS_SERVICE_ACCOUNT_KEY;
  if (!credentialsJson) {
    console.log("Could not parse GCS credentials from env var, trying file fallback...");
    return null;
  }
  
  try {
    const lastBraceIndex = credentialsJson.lastIndexOf('}');
    if (lastBraceIndex !== -1 && lastBraceIndex < credentialsJson.length - 1) {
      credentialsJson = credentialsJson.substring(0, lastBraceIndex + 1);
    }
    
    const credentials = JSON.parse(credentialsJson);
    const projectId = credentials.project_id;
    console.log(`[GCP] Credentials loaded from GCS_SERVICE_ACCOUNT_KEY for project: ${projectId}`);
    return { credentials, projectId };
  } catch (error: any) {
    console.error("[GCP] Failed to parse GCS_SERVICE_ACCOUNT_KEY:", error.message);
    return null;
  }
}

class GCPClientManager {
  private static instance: GCPClientManager;
  private storageClient: Storage | null = null;
  private kmsClient: KeyManagementServiceClient | null = null;
  private initialized = false;
  private initAttempted = false;
  private projectId: string | null = null;

  private constructor() {
  }
  
  private ensureInitialized(): void {
    if (!this.initAttempted) {
      this.initAttempted = true;
      this.initialize();
    }
  }

  private initialize(): void {
    try {
      const parsedCredentials = parseGCSCredentials();
      
      if (parsedCredentials) {
        this.projectId = parsedCredentials.projectId;
        const clientConfig = {
          projectId: parsedCredentials.projectId,
          credentials: parsedCredentials.credentials,
        };
        
        this.storageClient = new Storage(clientConfig);
        console.log("[GCP] Storage client initialized with service account credentials");

        if (isKMSConfigured()) {
          this.kmsClient = new KeyManagementServiceClient(clientConfig);
          console.log("[GCP] KMS client initialized");
        }

        this.initialized = true;
        console.log(`[GCP] All clients initialized for project: ${this.projectId}`);
        return;
      }
      
      if (process.env.GOOGLE_APPLICATION_CREDENTIALS || process.env.GCP_PROJECT_ID) {
        const clientConfig = GCP_CONFIG.PROJECT_ID ? { projectId: GCP_CONFIG.PROJECT_ID } : {};
        this.projectId = GCP_CONFIG.PROJECT_ID || null;

        this.storageClient = new Storage(clientConfig);
        console.log("[GCP] Storage client initialized with ADC");

        if (isKMSConfigured()) {
          this.kmsClient = new KeyManagementServiceClient(clientConfig);
          console.log("[GCP] KMS client initialized");
        }

        this.initialized = true;
        console.log(`[GCP] All clients initialized for project: ${this.projectId || "default"}`);
        return;
      }
      
      console.log("GCS credentials not available - GCS operations may fail");
    } catch (error: any) {
      console.error("[GCP] Failed to initialize clients:", error.message);
      this.initialized = false;
    }
  }

  public static getInstance(): GCPClientManager {
    if (!GCPClientManager.instance) {
      GCPClientManager.instance = new GCPClientManager();
    }
    return GCPClientManager.instance;
  }

  public getStorageClient(): Storage {
    this.ensureInitialized();
    if (!this.storageClient) {
      throw new Error("GCP Storage client not initialized. Check GCS_SERVICE_ACCOUNT_KEY or GOOGLE_APPLICATION_CREDENTIALS.");
    }
    return this.storageClient;
  }

  public getKMSClient(): KeyManagementServiceClient {
    this.ensureInitialized();
    if (!this.kmsClient) {
      throw new Error("GCP KMS client not initialized. Check GCP_KMS_* environment variables.");
    }
    return this.kmsClient;
  }

  public isInitialized(): boolean {
    this.ensureInitialized();
    return this.initialized;
  }

  public isStorageAvailable(): boolean {
    this.ensureInitialized();
    return this.storageClient !== null;
  }

  public isKMSAvailable(): boolean {
    this.ensureInitialized();
    return this.kmsClient !== null;
  }

  public getBucket(bucketName?: string) {
    this.ensureInitialized();
    const storage = this.getStorageClient();
    return storage.bucket(bucketName || GCP_CONFIG.STORAGE.BUCKET_NAME);
  }

  public getKMSKeyPath(): string {
    return `projects/${GCP_CONFIG.PROJECT_ID}/locations/${GCP_CONFIG.KMS.LOCATION}/keyRings/${GCP_CONFIG.KMS.KEY_RING}/cryptoKeys/${GCP_CONFIG.KMS.CRYPTO_KEY}`;
  }

  public getHealthcareDatasetPath(): string {
    return `projects/${GCP_CONFIG.PROJECT_ID}/locations/${GCP_CONFIG.HEALTHCARE.LOCATION}/datasets/${GCP_CONFIG.HEALTHCARE.DATASET_ID}`;
  }

  public getFHIRStorePath(): string {
    return `${this.getHealthcareDatasetPath()}/fhirStores/${GCP_CONFIG.HEALTHCARE.FHIR_STORE_ID}`;
  }

  public getDICOMStorePath(): string {
    return `${this.getHealthcareDatasetPath()}/dicomStores/${GCP_CONFIG.HEALTHCARE.DICOM_STORE_ID}`;
  }
}

export const gcpManager = GCPClientManager.getInstance();

export function getStorageClient(): Storage {
  return gcpManager.getStorageClient();
}

export function getKMSClient(): KeyManagementServiceClient {
  return gcpManager.getKMSClient();
}

export function isGCPAvailable(): boolean {
  return gcpManager.isInitialized();
}

export function isStorageAvailable(): boolean {
  return gcpManager.isStorageAvailable();
}

export function isKMSAvailable(): boolean {
  return gcpManager.isKMSAvailable();
}

export { GCP_CONFIG } from "./config/gcpConstants";

File: app/gcp_config.py
"""
Google Cloud Platform Client Initialization (Python)

Centralized initialization of all GCP clients for Python services.
Supports GCS_SERVICE_ACCOUNT_KEY env var (JSON string) or GOOGLE_APPLICATION_CREDENTIALS file.
"""

import os
import json
import logging
from typing import Optional, Dict, Any
from functools import lru_cache

from app.config.gcp_constants import GCP_CONFIG, is_gcp_configured, is_kms_configured

logger = logging.getLogger(__name__)

_storage_client = None
_kms_client = None
_initialized = False
_project_id = None


def _parse_gcs_credentials() -> Optional[Dict[str, Any]]:
    """Parse GCS credentials from GCS_SERVICE_ACCOUNT_KEY environment variable."""
    credentials_json = os.getenv("GCS_SERVICE_ACCOUNT_KEY")
    if not credentials_json:
        return None
    
    try:
        last_brace = credentials_json.rfind('}')
        if last_brace != -1 and last_brace < len(credentials_json) - 1:
            credentials_json = credentials_json[:last_brace + 1]
        
        credentials = json.loads(credentials_json)
        project_id = credentials.get("project_id")
        logger.info(f"[GCP] Credentials loaded from GCS_SERVICE_ACCOUNT_KEY for project: {project_id}")
        return credentials
    except json.JSONDecodeError as e:
        logger.error(f"[GCP] Failed to parse GCS_SERVICE_ACCOUNT_KEY: {e}")
        return None


def _initialize_clients():
    """Initialize GCP clients if credentials are available."""
    global _storage_client, _kms_client, _initialized, _project_id
    
    if _initialized:
        return
    
    parsed_credentials = _parse_gcs_credentials()
    
    if parsed_credentials:
        try:
            from google.cloud import storage
            from google.oauth2 import service_account
            
            _project_id = parsed_credentials.get("project_id")
            credentials = service_account.Credentials.from_service_account_info(parsed_credentials)
            
            _storage_client = storage.Client(project=_project_id, credentials=credentials)
            logger.info("[GCP] Storage client initialized with service account credentials")
            
            if is_kms_configured():
                try:
                    from google.cloud import kms
                    _kms_client = kms.KeyManagementServiceClient(credentials=credentials)
                    logger.info("[GCP] KMS client initialized")
                except Exception as e:
                    logger.error(f"[GCP] Failed to initialize KMS client: {e}")
            
            _initialized = True
            logger.info(f"[GCP] All clients initialized for project: {_project_id}")
            return
        except Exception as e:
            logger.error(f"[GCP] Failed to initialize with service account: {e}")
    
    credentials_path = os.getenv("GOOGLE_APPLICATION_CREDENTIALS")
    project_id = GCP_CONFIG.PROJECT_ID
    
    if not credentials_path and not project_id:
        logger.warning("[GCP] GOOGLE_APPLICATION_CREDENTIALS or GCP_PROJECT_ID not set. Using local fallback.")
        _initialized = True
        return
    
    try:
        from google.cloud import storage
        _storage_client = storage.Client(project=project_id if project_id else None)
        _project_id = project_id
        logger.info("[GCP] Storage client initialized with ADC")
    except Exception as e:
        logger.error(f"[GCP] Failed to initialize Storage client: {e}")
    
    if is_kms_configured():
        try:
            from google.cloud import kms
            _kms_client = kms.KeyManagementServiceClient()
            logger.info("[GCP] KMS client initialized")
        except Exception as e:
            logger.error(f"[GCP] Failed to initialize KMS client: {e}")
    
    _initialized = True
    logger.info(f"[GCP] Clients initialized for project: {project_id or 'default'}")


def get_storage_client():
    """Get the GCS storage client."""
    _initialize_clients()
    if _storage_client is None:
        raise RuntimeError("GCP Storage client not initialized. Check GOOGLE_APPLICATION_CREDENTIALS.")
    return _storage_client


def get_kms_client():
    """Get the Cloud KMS client."""
    _initialize_clients()
    if _kms_client is None:
        raise RuntimeError("GCP KMS client not initialized. Check GCP_KMS_* environment variables.")
    return _kms_client


def is_storage_available() -> bool:
    """Check if GCS storage is available."""
    _initialize_clients()
    return _storage_client is not None


def is_kms_available() -> bool:
    """Check if Cloud KMS is available."""
    _initialize_clients()
    return _kms_client is not None


def is_gcp_available() -> bool:
    """Check if any GCP services are available."""
    _initialize_clients()
    return _storage_client is not None or _kms_client is not None


@lru_cache(maxsize=1)
def get_bucket():
    """Get the default GCS bucket."""
    client = get_storage_client()
    return client.bucket(GCP_CONFIG.STORAGE.BUCKET_NAME)


def get_kms_key_path() -> str:
    """Get the full KMS key resource path."""
    return (
        f"projects/{GCP_CONFIG.PROJECT_ID}/locations/{GCP_CONFIG.KMS.LOCATION}/"
        f"keyRings/{GCP_CONFIG.KMS.KEY_RING}/cryptoKeys/{GCP_CONFIG.KMS.CRYPTO_KEY}"
    )
